osbuild/__init__.py

```diff
@@ -69,13 +70,156 @@ def __exit__(self, exc_type, exc_value, exc_tb):
+                # SELinux support is optional
+                if e.errno != errno.ENODATA:
+                    raise
+            else:
+                metadata["selinux"] = os.fsdecode(selinux_label)
+            # hash the JSON representation of the metadata to stay unique/stable/well-defined
+            m.update(json.dumps(metadata, sort_keys=True).encode())
+            if dirent.is_symlink():
+                m.update(os.fsdecode(os.readlink(dirent.name, dir_fd=dir_fd)).encode())
+            else:
+                fd = os.open(dirent.name, flags=os.O_RDONLY, dir_fd=dir_fd)
+                try:
+                    if dirent.is_dir(follow_symlinks=False):
+                        upsum(m, fd)
+                    elif dirent.is_file(follow_symlinks=False):
+                        # hash a page at a time (using f with fd as default is a hack to please pylint)
+                        for byte_block in iter(lambda f=fd: os.read(f, 4096), b""):
+                            m.update(byte_block)
+                    else:
+                        raise ValueError("Found unexpected filetype on OS image")
+                finally:
+                    os.close(fd)
+
+
+class ObjectStore:
+    def __init__(self, store):
+        objects = f"{store}/objects"
+        refs = f"{store}/refs"
+        os.makedirs(store, exist_ok=True)
+        os.makedirs(objects, exist_ok=True)
+        os.makedirs(refs, exist_ok=True)
+        self.store = store
+        self.objects = objects
+        self.refs = refs
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        pass
+
+    @contextlib.contextmanager
+    def get_tree(self, tree_id):
+        if not tree_id:
+            yield None
+        else:
+            with tempfile.TemporaryDirectory(dir=self.store) as tmp:
+                try:
+                    subprocess.run(["mount", "-o", "bind,ro,mode=0755", f"{self.refs}/{tree_id}", tmp], check=True)
+                except subprocess.CalledProcessError as e:
+                    if e.returncode == 32:
+                        # mount failed, we interpret this as tree_id not existing
+                        yield None
+                    else:
+                        raise
+                else:
+                    yield tmp
+                    subprocess.run(["umount", "--lazy", tmp], check=True)
+
+    @contextlib.contextmanager
+    def new_tree(self, tree_id=None, base_id=None):
+        with tempfile.TemporaryDirectory(dir=self.store) as tmp:
+            if not tree_id:
+                # no target tree is being generated
+                if base_id:
+                    # yield the base read-only, so we guarantee it is safe to
+                    # reuse this by others, but avoid having to make a copy
+                    subprocess.run(["mount", "-o", "bind,ro,mode=0755", f"{self.refs}/{base_id}", tmp], check=True)
+                    yield tmp
```

Comment:
Do we still need this case if we have `get_tree()`?

Reply:
Not any more, no (their semantics were different, but changed it).